File: tests/tensor/test_tensor.cpp
#include <iostream>
#include <vector>

#include <dl/dl.hpp>

int main() {
    dl::Tensor x = dl::Tensor::empty(
        dl::Shape({2, 3}),
        dl::DType::Float32,
        dl::Device(dl::DeviceType::CUDA, 0));
    dl::Tensor z = dl::Tensor::zeros(
        dl::Shape({2, 3}),
        dl::DType::Float32,
        dl::Device(dl::DeviceType::CUDA, 0));
    dl::Tensor x_like = dl::Tensor::empty_like(x);
    dl::Tensor z_like = dl::Tensor::zeros_like(x);
    const std::vector<float> zeros = z.to_host<float>();
    const std::vector<float> zeros_like = z_like.to_host<float>();

    bool zero_check = true;
    for (float value : zeros) {
        if (value != 0.0f) {
            zero_check = false;
            break;
        }
    }
    for (float value : zeros_like) {
        if (value != 0.0f) {
            zero_check = false;
            break;
        }
    }

    const bool passed = (x.numel() == 6 && x.nbytes() == 24 &&
                         x_like.numel() == x.numel() &&
                         x_like.dtype() == x.dtype() &&
                         x_like.device().type == x.device().type &&
                         x_like.device().index == x.device().index &&
                         zero_check);
    std::cout << "tensor_test : " << (passed ? "passed" : "not passed") << "\n";

    return passed ? 0 : 1;
}

File: tests/python/test_tensor.py
import pytest

import dl


def cuda_device():
    return dl.cuda(0)


def test_tensor_numel_nbytes():
    x = dl.Tensor((2, 3), dl.float32, cuda_device())

    assert x.numel() == 6
    assert x.nbytes() == 24
    assert x.shape.dims == [2, 3]
    assert x.dtype == dl.float32
    assert x.device.is_cuda()


def test_tensor_from_list_to_list():
    x = dl.Tensor.from_list([1.0, 2.0, 3.0, 4.0], (2, 2), cuda_device())

    assert x.tolist() == pytest.approx([1.0, 2.0, 3.0, 4.0])


def test_tensor_zeros():
    x = dl.Tensor.zeros((4,), dl.float32, cuda_device())

    assert x.tolist() == pytest.approx([0.0, 0.0, 0.0, 0.0])
